cloudinit/sources/DataSourceAzure.py

```diff
@@ -660,10 +691,249 @@ def setup(self, is_new_instance):
+
+    @azure_ds_telemetry_reporter
+    def _wait_for_hot_attached_nics(self, nl_sock):
+        """Wait until all the expected nics for the vm are hot-attached.
+        The expected nic count is obtained by requesting the network metadata
+        from IMDS.
+        """
+        LOG.info("Waiting for nics to be hot-attached")
+        try:
+            # Wait for nics to be attached one at a time, until we
+            # know for sure that all nics have been attached.
+            nics_found = []
+            primary_nic_found = False
+            expected_nic_count = -1
+
+            # Wait for netlink nic attach events.
+            # After the first nic is attached, we are already in the
+            # customer vm deployment path and so everything from then
+            # on should happen fast and avoid unnecessary delays
+            # wherever possible.
+            while True:
+                ifname = netlink.wait_for_nic_attach_event(nl_sock,
+                                                           nics_found)
+                nics_found.append(ifname)
+                LOG.info("Detected nic %s attached.", ifname)
+
+                # Attempt to bring the interface's operating state to
+                # UP in case it is not already.
+                self.wait_for_link_up(ifname)
+
+                # If primary nic is not found, check if this is it. The
+                # platform will attach the primary nic first so we
+                # won't be in primary_nic_found = false state for long.
+                if not primary_nic_found:
+                    LOG.info("Checking if %s is the primary nic",
+                             ifname)
+                    (primary_nic_found, expected_nic_count) = (
+                        self._check_if_nic_is_primary(ifname))
+
+                # Exit criteria: check if we've discovered all nics
+                if (expected_nic_count != -1
+                        and len(nics_found) >= expected_nic_count):
+                    LOG.info("Found all the nics for this VM.")
+                    break
+
+        except AssertionError as error:
+            report_diagnostic_event(error, logger_func=LOG.error)
+
+    @azure_ds_telemetry_reporter
+    def _wait_for_all_nics_ready(self):
+        """Wait for nic(s) to be hot-attached. There may be multiple nics
+           depending on the customer request.
+           But only primary nic would be able to communicate with wireserver
+           and IMDS. So we detect and save the primary nic to be used later.
+        """
+        # Report ready if the marker file is not already present.
+        # The nic of the preprovisioned vm gets hot-detached as soon as
+        # we report ready. So no need to save the dhcp context.
+        self._report_ready_if_needed()
+
+        has_nic_been_detached = bool(
+            os.path.isfile(REPROVISION_NIC_DETACHED_MARKER_FILE))
+
+        try:
+            nl_sock = netlink.create_bound_netlink_socket()
```

Comment:
As discussed offline:

Once report ready happens, the platform will detach the nic from the VM. If the nic is detached _before_ the netlink socket is created, then cloud-init will be stuck waiting for the nic detach event, as it has already happened.

The netlink socket should be created before reporting ready, so that the nic detach event is guaranteed to be captured by the netlink socket. This is so that after reporting ready, when we read the netlink messages, the nic detach event message can be consumed by cloud-init under all scenarios.

Once netlink socket creation is moved before reporting ready, please also add a comment about this for the information of future devs.